### PR TITLE
Add preliminary profiling support to the CLI

### PR DESCRIPTION
**Pull request by nalind**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add CPU and memory profiling to the CLI, as hidden options.  The picture they paint doesn't include anything we do in subprocesses, and we use a _lot_ of those when we're manipulating layers, but it can still be useful for analyzing our front-end logic.

#### How to verify it
Run any of our commands using the new global `--cpu-profile` and/or `--memory-profile` options, which take filenames as arguments, and then use `go tool pprof` to examine one of the generated files.

#### Which issue(s) this PR fixes:
None

#### Does this PR introduce a user-facing change?

```release-note
The new `--cpu-profile` and `--memory-profile` options are expected to be useful only for the maintainers, and are hidden by default.
```

### CODE REVIEW

**Comment by rhatdan** [comment]
@cevich Cirrus seems to be not running?

**Comment by nalind** [comment]
Looks like it caught up.

**Comment by rhatdan** [comment]
LGTM
@TomSweeneyRedHat PTAL

File: cmd/buildah/main.go [hunk 1]
```diff
@@ -27,6 +29,9 @@ type globalFlags struct {
 	StorageOpts       []string
 	UserNSUID         []string
 	UserNSGID         []string
+	CPUProfile        string
+	cpuProfile        *os.File
+	MemoryProfile     string
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/main.go:34]: not a huge fan of having the same variable names in a struct with only casing as a difference, but personal twitch.

Comment (rhatdan) [reply cmd/buildah/main.go:34]: Yes I agree, this is bad.  Can we change this to cpuProfileFD or something like that.

**Comment by TomSweeneyRedHat** [comment]
segv'd on me when using the memory-profile.  The cpu-profile seems to work fine, but when I also have --memory-profile, or it's by itself, I'm getting the below segv.

I think once settled, we ought to have a little paragraph or two in  CONTRIBUTING.md about this so we (I!) won't forget about it.

```
# cat ~/Dockerfile
# stable/Dockerfile
#
# Build a Buildah container image from the latest
# stable version of Buildah on the Fedoras Updates System.
# https://bodhi.fedoraproject.org/updates/?search=buildah
# This image can be used to create a secured container
# that runs safely with privileges within the container.
#
FROM fedora:latest

# Don't include container-selinux and remove
# directories used by yum that are just taking
# up space.
RUN useradd build; yum -y update; yum -y reinstall shadow-utils; yum -y install buildah fuse-overlayfs --exclude container-selinux; rm -rf /var/cache /var/log/dnf* /var/log/yum.*;

#ADD https://raw.githubusercontent.com/containers/buildah/master/contrib/buildahimage/stable/containers.conf /etc/containers/

# Adjust storage.conf to enable Fuse storage.
RUN sed -i -e 's|^#mount_program|mount_program|g' -e '/additionalimage.*/a "/var/lib/shared",' -e 's|^mountopt[[:space:]]*=.*$|mountopt = "nodev,fsync=0"|g' /etc/containers/storage.conf
RUN mkdir -p /var/lib/shared/overlay-images /var/lib/shared/overlay-layers; touch /var/lib/shared/overlay-images/images.lock; touch /var/lib/shared/overlay-layers/layers.lock

# Set an environment variable to default to chroot isolation for RUN
# instructions and "buildah run".
ENV BUILDAH_ISOLATION=chroot

########## Start of run ################

# buildah --cpu-profile /tmp/cpu.txt --memory-profile /tmp/mem.txt bud -t tom -f ~/Dockerfile .
STEP 1: FROM fedora:latest
STEP 2: RUN useradd build; yum -y update; yum -y reinstall shadow-utils; yum -y install buildah fuse-overlayfs --exclude container-selinux; rm -rf /var/cache /var/log/dnf* /var/log/yum.*;
Fedora Modular 31 - x86_64                                                                                                                             2.3 MB/s | 5.2 MB     00:02    

{Lot's of happy yum stuff removed for brevity}

Installed:
  buildah-1.14.3-1.fc31.x86_64     containers-common-1:0.1.41-1.fc31.x86_64     crun-0.13-1.fc31.x86_64     fuse-common-3.6.2-1.fc31.x86_64     fuse-overlayfs-0.7.8-1.fc31.x86_64    
  fuse3-3.6.2-1.fc31.x86_64        fuse3-libs-3.6.2-1.fc31.x86_64               kmod-26-4.fc31.x86_64       libslirp-4.1.0-1.fc31.x86_64        slirp4netns-1.0.0-1.fc31.x86_64       
  yajl-2.1.0-13.fc31.x86_64       

Complete!
STEP 3: RUN sed -i -e 's|^#mount_program|mount_program|g' -e '/additionalimage.*/a "/var/lib/shared",' -e 's|^mountopt[[:space:]]*=.*$|mountopt = "nodev,fsync=0"|g' /etc/containers/storage.conf
STEP 4: RUN mkdir -p /var/lib/shared/overlay-images /var/lib/shared/overlay-layers; touch /var/lib/shared/overlay-images/images.lock; touch /var/lib/shared/overlay-layers/layers.lock
STEP 5: ENV BUILDAH_ISOLATION=chroot
STEP 6: COMMIT tom
Getting image source signatures
Copying blob 1934db628f4c skipped: already exists  
Copying blob 797ba5bd62b5 done  
Copying config aceb1d7564 done  
Writing manifest to image destination
Storing signatures
--> aceb1d75644
aceb1d756442afb46036eec052a30e3e29576f37859c971f9b9f5392d9db7479
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1122b8a]

goroutine 1 [running]:
runtime/pprof.(*Profile).WriteTo(0x0, 0x166f560, 0xc0014e6000, 0x1, 0xc0014e6000, 0x0)
	/usr/lib/golang/src/runtime/pprof/pprof.go:325 +0x3a
main.after(0xc0004d1600, 0x0, 0x0)
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/cmd/buildah/main.go:164 +0x133
main.glob..func3(0xc0004d1600, 0xc0004e1a70, 0x1, 0x9, 0x0, 0x0)
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/cmd/buildah/main.go:47 +0x2b
github.com/spf13/cobra.(*Command).execute(0xc0004d1600, 0xc0004e19e0, 0x9, 0x9, 0xc0004d1600, 0xc0004e19e0)
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/vendor/github.com/spf13/cobra/command.go:853 +0x36f
github.com/spf13/cobra.(*Command).ExecuteC(0x20b8d00, 0x6, 0x14335e0, 0x8)
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/vendor/github.com/spf13/cobra/command.go:943 +0x318
github.com/spf13/cobra.(*Command).Execute(...)
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/vendor/github.com/spf13/cobra/command.go:883
main.main()
	/root/tsweeney/workspaces/containers/buildah/src/github.com/containers/buildah/cmd/buildah/main.go:179 +0xe3

```

**Comment by TomSweeneyRedHat** [comment]
LGTM and the segv has been chased away.  A follow up PR with doc added to CONTRIBUTING.md or some such would be appreciated.

**Comment by rhatdan** [comment]
bors retry

**Comment by TomSweeneyRedHat** [comment]
bors retry

**Comment by cevich** [comment]
hrmmm

**Comment by cevich** [comment]
[It's Ubuntu 19.04 that's failing](https://github.com/containers/buildah/runs/595433869)...uggg, another problem technically fixed by #1971 assuming I can get it to run :confused: 

TBH I'd be okay if you want to just disable that part of the matrix temporarily.  My other PR brings us up to the more current 19.10 anyway.

**Comment by cevich** [comment]
Optional "Fix" if desired: #2307

**Pull request merged**